include/dlaf/eigensolver/backtransformation/mc.h

```diff
@@ -93,125 +89,103 @@ void BackTransformation<Backend::MC, Device::CPU, T>::call_FC(
-                      std::move(tile_i0));
+                      std::move(tile_i_row));
       }
 
       // Copy VV into W
-      copySingleTile(mat_vv.read(LocalTileIndex(i, 0)), mat_w(LocalTileIndex(i, 0)));
+      copySingleTile(panelVV.read(i_row), panelW(i_row));
     }
 
-    // Reset W2 to zero
-    setZero(mat_w2);
-
-    // TODO: instead of using a full matrix, choose a "column" matrix. The problem is that last tile
-    // should be square but may have different size.
     const GlobalTileIndex v_start{k + 1, k};
     auto taus_panel = taus[k];
     const SizeType taupan = (is_last) ? last_mb : mat_v.blockSize().cols();
+    auto kk = LocalTileIndex{k, k};
+    panelT.setRange({k, k}, {k + 1, k + 1});
+    const LocalTileIndex diag_wp_idx{Coord::Row, k};
+    panelT.setWidth(dist_t.tileSize(GlobalTileIndex{k, k}).cols());
     dlaf::factorization::internal::computeTFactor<Backend::MC>(taupan, mat_v, v_start, taus_panel,
-                                                               mat_t(LocalTileIndex{k, k}));
+                                                               panelT(diag_wp_idx));
 
+    // WH = V T
     for (SizeType i = k + 1; i < m; ++i) {
-      auto kk = LocalTileIndex{k, k};
-      // WH = V T
-      auto ik = LocalTileIndex{i, 0};
-      hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile_t = mat_t.read(kk);
-      auto tile_w = mat_w(ik);
-
-      if (mat_t.tileSize(GlobalTileIndex{k, k}).rows() != mat_w.tileSize(GlobalTileIndex{i, 0}).cols()) {
-        TileElementIndex origin(0, 0);
-        TileElementSize size(mat_t.tileSize(GlobalTileIndex{k, k}).rows(),
-                             mat_t.tileSize(GlobalTileIndex{k, k}).cols());
-        const matrix::SubTileSpec spec({origin, size});
-        auto subtile_w = splitTile(tile_w, spec);
-        trmmPanel(executor_np, tile_t, std::move(subtile_w));
-      }
-      else
-        trmmPanel(executor_np, mat_t.read(kk), std::move(tile_w));
+      auto i_row = LocalTileIndex{Coord::Row, i};
+      hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile_t = panelT.read(diag_wp_idx);
+
+      panelW.setWidth(dist_t.tileSize(GlobalTileIndex{k, k}).cols());
+      trmmPanel(executor_np, tile_t, panelW(i_row));
     }
 
+    // W2 = W C
     for (SizeType j = 0; j < n; ++j) {
-      auto kj = LocalTileIndex{0, j};
+      auto j_col = LocalTileIndex{Coord::Col, j};
       for (SizeType i = k + 1; i < m; ++i) {
-        auto ik = LocalTileIndex{i, 0};
+        auto i_row = LocalTileIndex{Coord::Row, i};
         auto ij = LocalTileIndex{i, j};
-        // W2 = W C
-        gemmUpdateW2(executor_np, mat_w(ik), mat_c.read(ij), mat_w2(kj));
+        panelW.setWidth(mat_v.tileSize(GlobalTileIndex{i, k}).cols());
+        if ((i == k + 1)) {
+          gemmUpdateW2Start(executor_np, panelW(i_row), mat_c.read(ij), panelW2(j_col));
+        }
+        else {
+          gemmUpdateW2(executor_np, panelW(i_row), mat_c.read(ij), panelW2(j_col));
+        }

```

Comment:
I don't see the point of having two calls (forcing to find a good name for them), with the same parameters, that executes the exact same line of code, with just a single hard-coded change in the parameter.

Something similar to this?

```suggestion
        hpx::dataflow(ex, matrix::unwrapExtendTiles(tile::gemm_o), blas::Op::ConjTrans, blas::Op::NoTrans,
          T(1), panelW(i_row), mat_c.read(ij), (i == k + 1) ? T(0) : T(1), panelW2(j_col));
```

I agree that it may be useful for code-reusage (thinking about the distributed version of this), in that case I would suggest to evaluate to pass also the coefficient as parameter, or to group the entire W2 logic update into the function (e.g. `computeW2` instead of just doing a partial update of it).

I'm not really sure what's better, because the distributed version will require also a communication, so it would not be so direct the code re-usage.